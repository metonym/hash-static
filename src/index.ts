import chalk from 'chalk';
import fs from 'fs';
import htmlnano from 'htmlnano';
import posthtml from 'posthtml';
import { hash } from 'posthtml-hash';

export interface IHashStaticOptions {
  entry: string;
  minify?: boolean;
}

const DEFAULT_OPTIONS = {
  entry: 'dist/index.html',
  minify: true
};

async function hashStatic(options: IHashStaticOptions) {
  const entry = options.entry || DEFAULT_OPTIONS.entry;
  const minify =
    options && 'minify' in options ? options.minify : DEFAULT_OPTIONS.minify;

  if (!fs.existsSync(entry)) {
    throw Error('Entry does not exist.');
  }

  const [path] = entry.split('index.html');
  const plugins = [hash({ path })];

  if (minify) {
    plugins.push(htmlnano());
  }

  const html = fs.readFileSync(entry).toString();
  const result = await posthtml(plugins).process(html);

  fs.writeFileSync(entry, result.html);

  console.log(chalk.green('Successfully hashed static assets.'));
}

async function hashStaticCli(process: NodeJS.Process) {
  const flags = process.argv.slice(2);
  await hashStatic({ entry: flags[0] });
}

export default hashStatic;
export { hashStatic, hashStaticCli };
