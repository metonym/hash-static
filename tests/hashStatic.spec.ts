import fs from 'fs';
// tslint:disable-next-line: no-implicit-dependencies
import sh from 'shelljs';
import { hashStatic, hashStaticCli } from '../src';

describe('hashStatic', () => {
  beforeEach(() => {
    jest.spyOn(global.console, 'log');
    process.argv = process.argv.slice(0, 2);
  });

  afterEach(() => {
    jest.clearAllMocks();
    sh.rm('-rf', 'tests/fixtures/basic/processed');
  });

  it('throws an error if entry is invalid', () => {
    expect(
      hashStatic({ entry: 'tests/dist/index.html' })
    ).rejects.toMatchObject(
      "[TypeError: Cannot read property 'entry' of undefined]"
    );
  });

  test('hashStatic – matches snapshot (minified)', async () => {
    sh.cp(
      '-r',
      'tests/fixtures/basic/original',
      'tests/fixtures/basic/processed'
    );

    const entry = 'tests/fixtures/basic/processed/index.html';
    await hashStatic({ entry });
    const html = fs.readFileSync(entry).toString();
    expect(console.log).toBeCalledTimes(1);
    expect(html).toMatchSnapshot();
  });

  test('hashStatic – matches snapshot (non-minified)', async () => {
    sh.cp(
      '-r',
      'tests/fixtures/basic/original',
      'tests/fixtures/basic/processed'
    );

    const entry = 'tests/fixtures/basic/processed/index.html';
    await hashStatic({ entry, minify: false });
    const html = fs.readFileSync(entry).toString();
    expect(console.log).toBeCalledTimes(1);
    expect(html).toMatchSnapshot();
  });

  test('hashStaticCli – matches snapshot', async () => {
    sh.cp(
      '-r',
      'tests/fixtures/basic/original',
      'tests/fixtures/basic/processed'
    );

    const entry = 'tests/fixtures/basic/processed/index.html';
    process.argv.push(entry);
    await hashStaticCli(process);
    const html = fs.readFileSync(entry).toString();
    expect(console.log).toBeCalledTimes(1);
    expect(html).toMatchSnapshot();
  });
});
