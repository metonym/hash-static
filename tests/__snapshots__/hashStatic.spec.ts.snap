// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`hashStatic hashStatic – matches snapshot (minified) 1`] = `"<html><head><link rel=\\"stylesheet\\" href=\\"style.2d52e1da1da5cc437616.css\\"></head><body></body></html>"`;

exports[`hashStatic hashStatic – matches snapshot (non-minified) 1`] = `
"<html>
  <head>
    <link rel=\\"stylesheet\\" href=\\"style.2d52e1da1da5cc437616.css\\">
  </head>
  <body></body>
</html>
"
`;

exports[`hashStatic hashStaticCli – matches snapshot 1`] = `"<html><head><link rel=\\"stylesheet\\" href=\\"style.2d52e1da1da5cc437616.css\\"></head><body></body></html>"`;
